### ISSUE TITLE
Inference of image captioning on single image

### ISSUE THREAD

**Issue by ChandanVerma** [alibaba/AliceMind#78]
How do i used mPlug to generate image captions on a single image?

**Comment by huuquan1994** [comment]
@ChandanVerma 
Here is an example of how to load the mPLUG-Large-V2 model

### Load the mPLUG from checkpoint
```python
import torch.nn as nn
import yaml
from models.model_caption_mplug import MPLUG
from models.tokenization_bert import BertTokenizer
from models.vit import interpolate_pos_embed, resize_pos_embed

# Load the config file
config_path = 'AliceMind/mPLUG/configs/caption_mplug_large.yaml'
config = yaml.load(open(config_path, 'r'), Loader=yaml.Loader)

config["min_length"] = 8
config["max_length"] = 25
config["add_object"] = True
config["beam_size"] = 5
config['text_encoder'] = 'bert-base-uncased'
config['text_decoder'] = 'bert-base-uncased'

# initilize mPLUG-Large-V2 model
mplug_tokenizer = BertTokenizer.from_pretrained(config['text_encoder'])
mPLUG_model = MPLUG(config=config, tokenizer=mplug_tokenizer)

mPLUG_ckpt_path = '/path/to/mplug_large_v2.pth' # replace the path of your ckpt here
checkpoint = torch.load(mPLUG_ckpt_path, map_location='cpu')
state_dict = checkpoint['model']

# reshape positional embedding to accomodate for image resolution change
# ref: https://github.com/alibaba/AliceMind/blob/main/mPLUG/caption_mplug.py#L227
if config["clip_name"] == "ViT-B-16":
    num_patches = int(config["image_res"] * config["image_res"]/(16*16))
elif config["clip_name"] == "ViT-L-14":
   num_patches = int(config["image_res"] * config["image_res"]/(14*14))

pos_embed = nn.Parameter(torch.zeros(num_patches + 1, 768).float())

pos_embed = resize_pos_embed(state_dict['visual_encoder.visual.positional_embedding'].unsqueeze(0),
                                         pos_embed.unsqueeze(0))
state_dict['visual_encoder.visual.positional_embedding'] = pos_embed

# load and move the mPLUG-Large model to GPU
mPLUG_model.load_state_dict(state_dict, strict=False); # adding ; to avoid printing long messages
mPLUG_model = mPLUG_model.to('cuda').eval();
```

### Inference from single input
```python
from torchvision import transforms
from torchvision.transforms.functional import InterpolationMode

# define preprocess function
mPLUG_transform = transforms.Compose([
    transforms.Resize((config['image_res'], config['image_res']), 
                      interpolation=InterpolationMode.BICUBIC),
    transforms.ToTensor(),
    transforms.Normalize((0.48145466, 0.4578275, 0.40821073), 
                         (0.26862954, 0.26130258, 0.27577711)),
])

# load image
img_path = 'dog.jpg' # path to your image here
image = Image.open(img_path).convert('RGB')

# run inference
mplug_image = mPLUG_transform(image).unsqueeze(0) # [1, 3, image_res, image_res]
mplug_image = mplug_image.to('cuda')
    
topk_ids, topk_probs = mPLUG_model(mplug_image, None, train=False)
    
output_caption = mplug_tokenizer.decode(topk_ids[0][0]).replace("[SEP]", "").replace("[CLS]", "").replace("[PAD]", "").strip()

print(output_caption)
```

**Comment by ChandanVerma** [comment]
hi @huuquan1994 is it possible to generate video captions using a single video file as well? 
Thanks in advance.

**Comment by huuquan1994** [comment]
@ChandanVerma 
I don't have time to test on videos but you can try to refer to [https://github.com/alibaba/AliceMind/blob/main/mPLUG/videocap_mplug.py](https://github.com/alibaba/AliceMind/blob/main/mPLUG/videocap_mplug.py) and see how to load both videos and pre-trained model.
Take a look at the video dataset that uses to load video inputs (here: [https://github.com/alibaba/AliceMind/blob/main/mPLUG/dataset/video_dataset.py](https://github.com/alibaba/AliceMind/blob/main/mPLUG/dataset/video_dataset.py))

**Comment by ChandanVerma** [comment]
Sure.. thanks for the heads up :+1:

**Comment by CrazyBrick** [comment]
I got one.
`from models.model_caption_mplug_vatex import MPLUG
from models.vit import resize_pos_embed
from models.tokenization_bert import BertTokenizer

# import language_evaluation
from models.model_caption_mplug_vatex import MPLUG
from decord import VideoReader
device = 'cuda:0'
max_input_length = 30

# Load model

# model = model.to(device)
def initialize_model_videocap():
    config_path = 'configs/videocap_vatex_mplug_large.yaml'
    config = yaml.load(open(config_path, 'r'), Loader=yaml.Loader)
    config["min_length"] = 10
    config["max_length"] = 30
    config["beam_size"] = 5
    config['text_encoder'] = 'bert-base-uncased'
    config['text_decoder'] = 'bert-base-uncased'
    mplug_tokenizer = BertTokenizer.from_pretrained(config['text_encoder'])
    vdc_model = MPLUG(config=config, tokenizer=mplug_tokenizer)
    mPLUG_ckpt_path = 'checkpoints/mplug_large_v2.pth'
    checkpoint = torch.load(mPLUG_ckpt_path, map_location='cpu')
    state_dict = checkpoint['model']
    if config["clip_name"] == "ViT-B-16":
        num_patches = int(config["image_res"] * config["image_res"] / (16 * 16))
    elif config["clip_name"] == "ViT-L-14":
        num_patches = int(config["image_res"] * config["image_res"] / (14 * 14))

    pos_embed = nn.Parameter(torch.zeros(num_patches + 1, 768).float())

    pos_embed = resize_pos_embed(state_dict['visual_encoder.visual.positional_embedding'].unsqueeze(0),pos_embed.unsqueeze(0))
    state_dict['visual_encoder.visual.positional_embedding'] = pos_embed

    for key in list(state_dict.keys()):
        if ('fusion' in key or 'bert' in key) and 'decode' not in key:
            encoder_key = key.replace('fusion.', '').replace('bert.', '')
            state_dict[encoder_key] = state_dict[key]
            del state_dict[key]

    vdc_model.load_state_dict(state_dict, strict=False)
    vdc_model.cuda()
  
    return mplug_tokenizer, config, vdc_model

def load_video_from_path_decord(video_path, height, width,start_time=None, end_time=None, fps=-1):
    try:
        if not height or not width:
            vr = VideoReader(video_path)
        else:
            vr = VideoReader(video_path, width=width, height=height)

        vlen = len(vr)
        print(f"vlen: {vlen}")

        if start_time or end_time:
            assert fps > 0, 'must provide video fps if specifying start and end time.'

            start_idx = min(int(start_time * fps), vlen)
            end_idx = min(int(end_time * fps), vlen)
        else:
            start_idx, end_idx = 0, vlen

        num_frm = 4
        frame_indices = sorted(random.sample(range(vlen), num_frm))
        
        raw_sample_frms = vr.get_batch(frame_indices)
    except Exception as e:
        return None
    
    raw_sample_frms = raw_sample_frms.permute(0, 3, 1, 2)
    return raw_sample_frms

class ImageNorm(object):
    """Apply Normalization to Image Pixels on GPU
    """
    def __init__(self, mean, std):
        self.mean = torch.tensor(mean).view(1, 3, 1, 1)
        self.std = torch.tensor(std).view(1, 3, 1, 1)

    def __call__(self, img):
        if torch.max(img) > 1 and self.mean.max() <= 1:
            img.div_(255.)
        return img.sub_(self.mean).div_(self.std)
    
img_norm = ImageNorm(mean=(0.48145466, 0.4578275, 0.40821073), std=(0.26862954, 0.26130258, 0.27577711))

def open_video(video_fp):
    # load
    max_img_size = 384
    # vid_frm_array 中包含了整个视频的信息，每个维度对应于不同的属性。你可以使用它来访问视频的每一帧以及每一帧中的像素数据
    vid_frm_array  = load_video_from_path_decord(video_fp, height=max_img_size, width=max_img_size)
    # norm
    video = img_norm(vid_frm_array.float())
    
    return video
# input and generate
def generate_vdc(mplug_tokenizer, config, vdc_model,video_path):
    video = open_video(video_path)
    # mplug_vdc = mPLUG_transform(video).unsqueeze(0)
    mplug_vdc = video.unsqueeze(0)
    mplug_vdc = mplug_vdc.to('cuda')

    if config['prompt'] != "":
        caption = [config['prompt'] + config['eos']] * mplug_vdc.size(0)
        caption = mplug_tokenizer(caption, padding='longest', truncation=True, max_length=max_input_length,return_tensors="pt").to(device)
    else:
        caption = None
    print(f"caption: {caption}")
 
    topk_ids, topk_probs = vdc_model(mplug_vdc, caption, None, train=False, device=device)
    print(topk_probs)
    ans = mplug_tokenizer.decode(topk_ids[0][0]).replace("[SEP]", "").replace("[CLS]", "").replace("[PAD]", "").strip()
    ans += ' .'
    print(f"ans: {ans}")


video_path = 'xxx.mp4'
mplug_tokenizer, config, vdc_model = initialize_model_videocap()
generate_vdc(mplug_tokenizer, config, vdc_model,video_path)`

**Comment by CrazyBrick** [comment]
I got one.

from models.model_caption_mplug_vatex import MPLUG
from models.vit import resize_pos_embed
from models.tokenization_bert import BertTokenizer

# import language_evaluation
from models.model_caption_mplug_vatex import MPLUG
from decord import VideoReader
device = 'cuda:0'
max_input_length = 30

# Load model

# model = model.to(device)
def initialize_model_videocap():
    config_path = 'configs/videocap_vatex_mplug_large.yaml'
    config = yaml.load(open(config_path, 'r'), Loader=yaml.Loader)
    config["min_length"] = 10
    config["max_length"] = 30
    config["beam_size"] = 5
    config['text_encoder'] = 'bert-base-uncased'
    config['text_decoder'] = 'bert-base-uncased'
    mplug_tokenizer = BertTokenizer.from_pretrained(config['text_encoder'])
    vdc_model = MPLUG(config=config, tokenizer=mplug_tokenizer)
    mPLUG_ckpt_path = 'checkpoints/mplug_large_v2.pth'
    checkpoint = torch.load(mPLUG_ckpt_path, map_location='cpu')
    state_dict = checkpoint['model']
    if config["clip_name"] == "ViT-B-16":
        num_patches = int(config["image_res"] * config["image_res"] / (16 * 16))
    elif config["clip_name"] == "ViT-L-14":
        num_patches = int(config["image_res"] * config["image_res"] / (14 * 14))

    pos_embed = nn.Parameter(torch.zeros(num_patches + 1, 768).float())

    pos_embed = resize_pos_embed(state_dict['visual_encoder.visual.positional_embedding'].unsqueeze(0),pos_embed.unsqueeze(0))
    state_dict['visual_encoder.visual.positional_embedding'] = pos_embed

    for key in list(state_dict.keys()):
        if ('fusion' in key or 'bert' in key) and 'decode' not in key:
            encoder_key = key.replace('fusion.', '').replace('bert.', '')
            state_dict[encoder_key] = state_dict[key]
            del state_dict[key]

    vdc_model.load_state_dict(state_dict, strict=False)
    vdc_model.cuda()
  
    return mplug_tokenizer, config, vdc_model

def load_video_from_path_decord(video_path, height, width,start_time=None, end_time=None, fps=-1):
    try:
        if not height or not width:
            vr = VideoReader(video_path)
        else:
            vr = VideoReader(video_path, width=width, height=height)

        vlen = len(vr)
        print(f"vlen: {vlen}")

        if start_time or end_time:
            assert fps > 0, 'must provide video fps if specifying start and end time.'

            start_idx = min(int(start_time * fps), vlen)
            end_idx = min(int(end_time * fps), vlen)
        else:
            start_idx, end_idx = 0, vlen

        num_frm = 4
        frame_indices = sorted(random.sample(range(vlen), num_frm))
        
        raw_sample_frms = vr.get_batch(frame_indices)
    except Exception as e:
        return None
    
    raw_sample_frms = raw_sample_frms.permute(0, 3, 1, 2)
    return raw_sample_frms

class ImageNorm(object):
    """Apply Normalization to Image Pixels on GPU
    """
    def __init__(self, mean, std):
        self.mean = torch.tensor(mean).view(1, 3, 1, 1)
        self.std = torch.tensor(std).view(1, 3, 1, 1)

    def __call__(self, img):
        if torch.max(img) > 1 and self.mean.max() <= 1:
            img.div_(255.)
        return img.sub_(self.mean).div_(self.std)
    
img_norm = ImageNorm(mean=(0.48145466, 0.4578275, 0.40821073), std=(0.26862954, 0.26130258, 0.27577711))

def open_video(video_fp):
    # load
    max_img_size = 384
    # vid_frm_array 中包含了整个视频的信息，每个维度对应于不同的属性。你可以使用它来访问视频的每一帧以及每一帧中的像素数据
    vid_frm_array  = load_video_from_path_decord(video_fp, height=max_img_size, width=max_img_size)
    # norm
    video = img_norm(vid_frm_array.float())
    
    return video
# input and generate
def generate_vdc(mplug_tokenizer, config, vdc_model,video_path):
    video = open_video(video_path)
    # mplug_vdc = mPLUG_transform(video).unsqueeze(0)
    mplug_vdc = video.unsqueeze(0)
    mplug_vdc = mplug_vdc.to('cuda')

    if config['prompt'] != "":
        caption = [config['prompt'] + config['eos']] * mplug_vdc.size(0)
        caption = mplug_tokenizer(caption, padding='longest', truncation=True, max_length=max_input_length,return_tensors="pt").to(device)
    else:
        caption = None
    print(f"caption: {caption}")
 
    topk_ids, topk_probs = vdc_model(mplug_vdc, caption, None, train=False, device=device)
    print(topk_probs)
    ans = mplug_tokenizer.decode(topk_ids[0][0]).replace("[SEP]", "").replace("[CLS]", "").replace("[PAD]", "").strip()
    ans += ' .'
    print(f"ans: {ans}")


video_path = 'xxx.mp4'
mplug_tokenizer, config, vdc_model = initialize_model_videocap()
generate_vdc(mplug_tokenizer, config, vdc_model,video_path)
